ls1intum/Artemis
# `Notifications:` Improve behavior of exercise release date updates and scheduled notifications ⏰🔔

#### General
#### Server

### Motivation and Context
@lschlesinger noticed the following bug:
You create a new exercise with DueDate X. 
X is in the future (e.g. after 15 minutes)
You update the DueDate of the Exercise and set it to Z.
Z is also in the future but sooner. (e.g. only after 5 minutes)

Bug: 
- No notification was sent at the time of Z
(The scheduled notification for the time of X was correctly caught and removed)

Solution: 
When an exercise is Updated I now also update the scheduling process for notifications.

This PR also adds a new testing class for GroupNotificationService with the newly added functions tested 100%
+ I refactored the scattered copy&paste of the notification scheduling to the GroupNotificationService, therefore resolving the copy&paste antipattern. (Previously I could not figure out how to resolve this issue due to cyclic dependencies but now I simply have the class that lead to the cyclic dependency as an method parameter and no longer as part of the constructor for GroupNotificationService, therefore omitting the cyclic issues)

The issue occurred because scheduled notifications were only created in the notification pipeline for the notification type EXERCISE_RELEASED but not for the type EXERCISE_UPDATED.

### Steps for Testing
1) Create a new exercise with a release date in e.g. 5 minutes (the difference between the new and old release date has to be more than 2 minutes otherwise it will not work!)
2) Update this exercise due date and set it to e.g. 2 minutes in the future
3) Wait (e.g. 2 minutes) and verify that you received a notification.
4) Wait (e.g. 5 minutes) and verify that you did not receive any notification.

5) Feel free to test the release/update of any/all exercise types.

### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage
This PR also adds a new testing class for GroupNotificationService with the newly added functions tested 100%

The other files were not really changed (i.e. the distributed logic  is now bundled in a common point -> GroupNotificationService)

Locally all tests pass
![image](https://user-images.githubusercontent.com/65814168/138105771-f3180251-c4b8-4a9d-a8a2-6d0b1aee4637.png)
